doc/ocpasswd.8.md
# ocpasswd(8) -- OpenConnect server password utility

## SYNOPSIS
**ocpasswd** [--option-name[=value]] ['username']


## DESCRIPTION
This  program is openconnect password (ocpasswd) utility. It allows the generation
and handling of a 'plain' password file used by ocserv.

## OPTIONS

  * **-c, --passwd**=_FILE_::
    Specify the password file to use. Only useful when not using the default
    location.

  * **-g, --group**::
    Specify the user's group name. Note that groups of one character such as
    '*' and 'x', are ignored.

  * **-d, --delete**::
    Deletes the specified user from the password file.

  * **-l, --lock**::
    Prevents the specified user from logging in by locking its password.

  * **-u, --unlock**::
    Re-enables login for the specified user by unlocking its password.

  * **-h, --help**::
    Display usage information and exit.

  * **-v, --version**::
    Output version of program and exit.

## FILES
The password format of ocpasswd is as follows.

```
username:groupname:encoded-password
```

The crypt(3) encoding is used for the encoded-password.

## EXAMPLES

### Adding a user

```
$ ocpasswd -c ocpasswd my_username
```

### Locking a user

```
$ ocpasswd -c ocpasswd -l my_username
```

### Unlocking a user

```
$ ocpasswd -c ocpasswd -u my_username
```

## Exit status

  * **0**:
    Successful program execution.

  * **1**:
    The operation failed or the command syntax was not valid.

## SEE ALSO

ocserv(8), occtl(8)

## COPYRIGHT
Copyright (C) 2013-2017 Nikos Mavrogiannopoulos and others, all rights reserved.
This program is released under the terms of the GNU General Public License, version 2.

## AUTHORS
Written by Nikos Mavrogiannopoulos. Many people have
contributed to it.



NEWS
* Version 1.1.1 (released 2020-09-21)
- Fixed compatibility with OpenBSD that lacks procfs (#312)
- Improved rate-limit-ms and made it dependent on secmod backlog. This
  makes the server more resilient (and prevents connection failures) on
  multiple concurrent connections (#310)
- Added namespace support for listen address by introducing the
  listen-netns  option (#316)
- Disable TLS1.3 when cisco client compatibility is enabled. New
  anyconnect clients seem to supporting TLS1.3 but are unable to handle
  a client with an RSA key (#318)
- Enable a race free user disconnection via occtl (#59)
- Added the config option of a pre-login-banner (#313)
- Ocserv siwtched to using multiple ocserv-sm processes to improve scale,
  with the number of ocserv-sm process dependent on maximum clients and
  number of CPUs. Configuration option sec-mod-scale can be used to override
  the heuristics.
- Fixed issue with group selection on radius servers sending multiple
  group class attribute (#322).


* Version 1.1.0 (released 2020-06-16)
- Switch from fork to fork/exec model to achieve better scaling and ASLR
  protection. This introduces an ocserv-worker application which should be
  installed at the same path as ocserv (#285).
- When Linux OOM takes control kill ocserv workers before ocserv-main
  or ocserv-secmod (#283).
- Disable TCP queuing on the TLS port.
- Fix leak of GnuTLS session when DTLS connection is re-established (#293).


* Version 1.0.1 (released 2020-04-09)
- Prevent clients that use broken versions of gnutls from
  connecting using DTLS (#277).
- occtl: added machine-readable fields in json output (#271).
- occtl: IPs in ban list value is now reflecting the actual
  banned IPs rather than the database size (#272).


* Version 1.0.0 (released 2020-03-20)
- Avoid crash on invalid configuration values.
- Updated manpage generation to work with newer versions of ronn.
- Ensure scripts have all the information on all disconnection types (#231)
- Several updates to further restrict the control that worker processes have
  on the main process. Patches by Alan Jowett (#251, #252, #245, #247, #246)
- Add support for RFC6750 bearer tokens. This adds the "auth=oidc" config
  option. See doc/README-oidc.md for more information (#240).
- Add USER_AGENT, DEVICE_TYPE and DEVICE_PLATFORM environment variables when
  connect/disconnect scripts execute.
- Corrected issue with DTLS-PSK negotiation which prevented it from being
  enabled.
- Improved IPv6 handling of AnyConnect client for Apple ios (#254).
- Fixed issue with Radius accounting (#269).


* Version 0.12.6 (released 2019-12-28)
- Improved IPv6 support for anyconnect clients. Patch by Leendert van Doorn.
- The 'split-dns' configuration directive can be used per-user (#229).
- The max-same-clients=1 configuration option no longer refuses the
  reconnection of an already connected user (#223).
- Added openat() to the accepted list of seccomp calls. This allows ocserv
  to run under certain libcs (#185).


* Version 0.12.5 (released 2019-10-16)
- Fixed issue with FreeBSD tun devices closing (#213).
- Added configuration option udp-listen-host. This option supports different
  listen addresses for tcp and udp such as haproxy for tcp, but support dtls
  at the same time.
- occtl: fixed json output of show status command. Introduced tests
  for checking its json output using yajl (#220).
- occtl: use maxminddb when available.


* Version 0.12.4 (released 2019-07-03)
- Added support for radius access-challenge (multifactor) authentication.
- Fixed race condition when connect-script and disconnect-script are
  set, which could potentially cause a crash (#208).
- Perform quicker cleanup of sessions which their user explicitly
  disconnected (#210).


* Version 0.12.3 (released 2019-03-12)
- Fixed crash when no DTLS ciphersuite is negotiated.
- Fixed crash happening arbitrarily depending on handled string
  sizes (#197).
- Fixed compatibility issue with GnuTLS 3.3.x (#201).
- occtl: print the TLS session information, even if the DTLS channel
  is not established.


* Version 0.12.2 (released 2019-01-10)
- Added support for AES256-SHA legacy cipher. This allows the anyconnect
  clients to use AES256.
- Added support for the DTLS1.2 protocol hack used by new Anyconnect clients.


* Version 0.12.1 (released 2018-05-12)
- Fixed crash on initialization when server was running on background (#154)
- Work around issues with GnuTLS 3.4.x on ubuntu 16.04, at the cost
  of a memory leak on key reload (#152)


* Version 0.12.0 (released 2018-04-22)
- Allow DTLS stream to come from different IP from TLS stream.
  There are situations where internet providers send the UDP 
  stream from different IP (#61).
- Increased possibilities of allowed combinations of authentication
  methods (#108).
- Corrected regression since 0.11.8 with OTP authentication (#137).
- Added support for hostname-based virtual hosts, utilizing TLS
  SNI. With that change it is possible to configure multiple servers
  running over the same port (#133).
- Rename the tun device on BSD systems which support SIOCSIFNAME
  ioctl.
- Correctly handle proxy-protocol's health commands. That eliminates
  few connection drops when proxy protocol is in use.
- Corrected crash on certain cases when proxy protocol is in use (#146).


* Version 0.11.10 (released 2018-01-07)
- Increased the DTLS handshake timeout to 60 seconds and decreased
  retransmission time to 400ms.
- Improved compatibility with certain anyconnect clients which
  disconnect and reconnect after session establishment. Sessions
  which are explicitly terminated by the user, remain active for
  few seconds (15) to allow for anyconnect clients' reconnections.
- Improved cookie behavior on client timeout. The cookies remain
  active for their validity time, allowing mobile clients to get
  into sleep and recover session using the same cookie.
- Addressed issue which could cause the security module enter an

src/ocpasswd/ocpasswd.c
/*
 * Copyright (C) 2013 Nikos Mavrogiannopoulos
 *
 * This file is part of ocserv.
 *
 * ocserv is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * ocserv is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#ifndef _XOPEN_SOURCE
# define _XOPEN_SOURCE
#endif
#include <unistd.h>
#include <gnutls/gnutls.h>
#include <gnutls/crypto.h>	/* for random */
#include <sys/types.h>
#include <sys/stat.h>
#include <getopt.h>
#ifdef HAVE_CRYPT_H
  /* libcrypt in Fedora28 does not provide prototype
   * in unistd.h */
# include <crypt.h>
#endif

/* Gnulib portability files. */
#include <getpass.h>
#include <minmax.h>

#define DEFAULT_OCPASSWD "/etc/ocserv/ocpasswd"

static const char alphabet[] =
	"0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ./";

#define SALT_SIZE 16
static void
crypt_int(const char *fpasswd, const char *username, const char *groupname,
	  const char *passwd)
{
	uint8_t _salt[SALT_SIZE];
	char salt[SALT_SIZE+16];
	char *p, *cr_passwd;
	char *tmp_passwd;
	unsigned i;
	unsigned fpasswd_len = strlen(fpasswd);
	unsigned tmp_passwd_len;
	unsigned username_len = strlen(username);
	struct stat st;
	FILE *fd, *fd2;
	char *line = NULL;
	size_t line_size;
	ssize_t len, l;
	int ret;

	ret = gnutls_rnd(GNUTLS_RND_NONCE, _salt, sizeof(_salt));
	if (ret < 0) {
		fprintf(stderr, "Error generating nonce: %s\n",
			gnutls_strerror(ret));
		exit(1);
	}

#ifdef TRY_SHA2_CRYPT
	strcpy(salt, "$5$");
#else
	strcpy(salt, "$1$");
#endif
	p = salt + 3;

	for (i = 0; i < sizeof(_salt); i++) {
		*p = alphabet[_salt[i] % (sizeof(alphabet) - 1)];
		p++;
	}
	*p = '$';
	p++;
	*p = 0;
	p++;

	cr_passwd = crypt(passwd, salt);
	if (cr_passwd == NULL) { /* try MD5 */
		salt[1] = '1';
		cr_passwd = crypt(passwd, salt);
	}
	if (cr_passwd == NULL) {
		fprintf(stderr, "Error in crypt().\n");
		exit(1);
	}

	tmp_passwd_len = fpasswd_len + 5;
	tmp_passwd = malloc(tmp_passwd_len);
	if (tmp_passwd == NULL) {
		fprintf(stderr, "memory error\n");
		exit(1);
	}

	snprintf(tmp_passwd, tmp_passwd_len, "%s.tmp", fpasswd);
	if (stat(tmp_passwd, &st) != -1) {
		fprintf(stderr, "file '%s' is locked.\n", fpasswd);
		exit(1);
	}

	fd2 = fopen(tmp_passwd, "w");
	if (fd2 == NULL) {
		fprintf(stderr, "Cannot open '%s' for writing.\n", tmp_passwd);
		exit(1);
	}

	fd = fopen(fpasswd, "r");
	if (fd == NULL) {
		fprintf(fd2, "%s:%s:%s\n", username, groupname, cr_passwd);
	} else {
		int found = 0;
		while ((len = getline(&line, &line_size, fd)) > 0) {
			p = strchr(line, ':');
			if (p == NULL)
				continue;

			l = p-line;
			if (l == username_len && strncmp(line, username, l) == 0) {
				fprintf(fd2, "%s:%s:%s\n", username, groupname, cr_passwd);
				found = 1;
			} else {
				fwrite(line, 1, len, fd2);
			}
		}
		free(line);
		fclose(fd);

		if (found == 0)
			fprintf(fd2, "%s:%s:%s\n", username, groupname, cr_passwd);
	}

	fclose(fd2);

	ret = rename(tmp_passwd, fpasswd);
	if (ret < 0) {
		fprintf(stderr, "Cannot write to '%s'.\n", fpasswd);
		exit(1);
	}
	free(tmp_passwd);
}

static void
delete_user(const char *fpasswd, const char *username)
{
	FILE * fd, *fd2;
	char *tmp_passwd;
	char *line, *p;
	unsigned fpasswd_len = strlen(fpasswd);
	unsigned tmp_passwd_len;
	unsigned username_len = strlen(username);
	int ret;
	ssize_t len, l;
	size_t line_size;
	struct stat st;

	tmp_passwd_len = fpasswd_len + 5;
	tmp_passwd = malloc(tmp_passwd_len);

	snprintf(tmp_passwd, tmp_passwd_len, "%s.tmp", fpasswd);
	if (stat(tmp_passwd, &st) != -1) {
		fprintf(stderr, "file '%s' is locked.\n", fpasswd);
		exit(1);
	}

	fd = fopen(fpasswd, "r");
	if (fd == NULL) {
		fprintf(stderr, "Cannot open '%s' for reading.\n", fpasswd);
		exit(1);
	}

	fd2 = fopen(tmp_passwd, "w");
	if (fd2 == NULL) {
		fprintf(stderr, "Cannot open '%s' for writing.\n", tmp_passwd);
		exit(1);
	}

	line = NULL;
	while ((len = getline(&line, &line_size, fd)) > 0) {
		p = strchr(line, ':');
		if (p == NULL)
			continue;

		l = p-line;
		if (l == username_len && strncmp(line, username, l) == 0) {
			continue;
		} else {
			fwrite(line, 1, len, fd2);
		}
	}

	free(line);
	fclose(fd);
	fclose(fd2);

	ret = rename(tmp_passwd, fpasswd);
	if (ret == -1) {
		fprintf(stderr, "Cannot write to '%s'.\n", fpasswd);
		exit(1);
	}
	free(tmp_passwd);
}

static void
lock_user(const char *fpasswd, const char *username)
{
	FILE * fd, *fd2;
	char *tmp_passwd;
	char *line, *p;
	unsigned fpasswd_len = strlen(fpasswd);
	unsigned tmp_passwd_len;
	unsigned username_len = strlen(username);
	int ret;
	ssize_t len, l;
	size_t line_size;
	struct stat st;

	tmp_passwd_len = fpasswd_len + 5;
	tmp_passwd = malloc(tmp_passwd_len);

	snprintf(tmp_passwd, tmp_passwd_len, "%s.tmp", fpasswd);
	if (stat(tmp_passwd, &st) != -1) {
		fprintf(stderr, "file '%s' is locked.\n", fpasswd);
		exit(1);
	}

	fd = fopen(fpasswd, "r");
	if (fd == NULL) {
		fprintf(stderr, "Cannot open '%s' for reading.\n", fpasswd);
		exit(1);
	}

	fd2 = fopen(tmp_passwd, "w");
	if (fd2 == NULL) {
		fprintf(stderr, "Cannot open '%s' for writing.\n", tmp_passwd);
		exit(1);
	}

	line = NULL;
	while ((len = getline(&line, &line_size, fd)) > 0) {
		p = strchr(line, ':');
		if (p == NULL)
			continue;

		l = p-line;
		if (l == username_len && strncmp(line, username, l) == 0) {
			p = strchr(p+1, ':');
			if (p == NULL)
				continue;
			p++;

			if(*p != '!') {
				l = p-line;
				fwrite(line, 1, l, fd2);
				fputc('!', fd2);
				fwrite(p, 1, len-l, fd2);
			} else {
				fwrite(line, 1, len, fd2);
			}
		} else {
			fwrite(line, 1, len, fd2);
		}
	}

	free(line);
	fclose(fd);
	fclose(fd2);

	ret = rename(tmp_passwd, fpasswd);
	if (ret == -1) {
		fprintf(stderr, "Cannot write to '%s'.\n", fpasswd);
		exit(1);
	}
	free(tmp_passwd);
}

static void
unlock_user(const char *fpasswd, const char *username)
{
	FILE * fd, *fd2;
	char *tmp_passwd;
	char *line, *p;
	unsigned fpasswd_len = strlen(fpasswd);
	unsigned tmp_passwd_len;
	unsigned username_len = strlen(username);
	int ret;
	ssize_t len, l;
	size_t line_size;
	struct stat st;

	tmp_passwd_len = fpasswd_len + 5;
	tmp_passwd = malloc(tmp_passwd_len);

	snprintf(tmp_passwd, tmp_passwd_len, "%s.tmp", fpasswd);
	if (stat(tmp_passwd, &st) != -1) {
		fprintf(stderr, "file '%s' is locked.\n", fpasswd);
		exit(1);
	}

	fd = fopen(fpasswd, "r");
	if (fd == NULL) {
		fprintf(stderr, "Cannot open '%s' for reading.\n", fpasswd);
		exit(1);
	}

	fd2 = fopen(tmp_passwd, "w");
	if (fd2 == NULL) {
		fprintf(stderr, "Cannot open '%s' for writing.\n", tmp_passwd);
		exit(1);
	}

	line = NULL;
	while ((len = getline(&line, &line_size, fd)) > 0) {
		p = strchr(line, ':');
		if (p == NULL)
			continue;

		l = p-line;
		if (l == username_len && strncmp(line, username, l) == 0) {
			p = strchr(p+1, ':');
			if (p == NULL)
				continue;
			p++;

			l = p-line;
			fwrite(line, 1, l, fd2);

			if (*p=='!') p++;
			l = p-line;
			fwrite(p, 1, len-l, fd2);
		} else {
			fwrite(line, 1, len, fd2);
		}
	}

	free(line);
	fclose(fd);
	fclose(fd2);

	ret = rename(tmp_passwd, fpasswd);
	if (ret == -1) {
		fprintf(stderr, "Cannot write to '%s'.\n", fpasswd);
		exit(1);
	}
	free(tmp_passwd);
}

static const struct option long_options[] = {
	{"passwd", 1, 0, 'c'},
	{"groupname", 1, 0, 'g'},
	{"delete", 0, 0, 'd'},
	{"lock", 0, 0, 'l'},
	{"unlock", 0, 0, 'u'},
	{"help", 0, 0, 'h'},
	{"version", 0, 0, 'v'},
	{NULL, 0, 0, 0}
};

static
void usage(void)
{
	fprintf(stderr, "ocpasswd - OpenConnect server password utility\n");
	fprintf(stderr, "Usage:  ocpasswd [ -<flag> [<val>] | --<name>[{=| }<val>] ]... [username]\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "   -c, --passwd=file          Password file\n");
	fprintf(stderr, "   -g, --groupname=str        User's group name\n");
	fprintf(stderr, "   -d, --delete               Delete user\n");
	fprintf(stderr, "   -l, --lock                 Lock user\n");
	fprintf(stderr, "   -u, --unlock               Unlock user\n");
	fprintf(stderr, "   -v, --version              output version information and exit\n");
	fprintf(stderr, "   -h, --help                 display extended usage information and exit\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Options are specified by doubled hyphens and their name or by a single\n");
	fprintf(stderr, "hyphen and the flag character.\n\n");
	fprintf(stderr, "This program is openconnect password (ocpasswd) utility.  It allows the\n");
	fprintf(stderr, "generation and handling of a 'plain' password file used by ocserv.\n\n");
	fprintf(stderr, "Please send bug reports to:  "PACKAGE_BUGREPORT"\n");
}

static
void version(void)
{
	fprintf(stderr, "ocpasswd - "VERSION"\n");
	fprintf(stderr, "Copyright (C) 2013-2017 Nikos Mavrogiannopoulos, all rights reserved.\n");
	fprintf(stderr, "This is free software. It is licensed for use, modification and\n");
	fprintf(stderr, "redistribution under the terms of the GNU General Public License,\n");
	fprintf(stderr, "version 2 <http://gnu.org/licenses/gpl.html>\n\n");
	fprintf(stderr, "Please send bug reports to:  "PACKAGE_BUGREPORT"\n");
}

#define FLAG_DELETE 1
#define FLAG_LOCK (1<<1)
#define FLAG_UNLOCK (1<<2)

int main(int argc, char **argv)
{
	int ret, c;
	const char *username = NULL;
	char *groupname = NULL, *fpasswd = NULL;
	char* passwd = NULL;
	unsigned free_passwd = 0;
	size_t l, i;
	unsigned flags = 0;

	if ((ret = gnutls_global_init()) < 0) {
		fprintf(stderr, "global_init: %s\n", gnutls_strerror(ret));
		exit(1);
	}

	umask(066);

	while (1) {
		c = getopt_long(argc, argv, "c:g:dluvh", long_options, NULL);
		if (c == -1)
			break;

		switch(c) {
			case 'c':
				if (fpasswd) {
					fprintf(stderr, "-c option cannot be specified multiple time\n");
					exit(1);
				}
				fpasswd = strdup(optarg);
				break;
			case 'g':
				if (groupname) {
					fprintf(stderr, "-g option cannot be specified multiple time\n");
					exit(1);
				}
				groupname = strdup(optarg);
				break;
			case 'd':
				if (flags) {
					usage();
					exit(1);
				}
				flags |= FLAG_DELETE;
				break;
			case 'u':
				if (flags) {
					usage();
					exit(1);
				}
				flags |= FLAG_UNLOCK;
				break;
			case 'l':
				if (flags) {
					usage();
					exit(1);
				}
				flags |= FLAG_LOCK;
				break;
			case 'h':
				usage();
				exit(0);
			case 'v':
				version();
				exit(0);
		}
	}

	if (optind < argc && argc-optind == 1) {
		username = argv[optind++];
	} else {
		usage();
		exit(1);
	}

	if (!groupname)
		groupname = strdup("*");
	if (!fpasswd)
		fpasswd = strdup(DEFAULT_OCPASSWD);

	if (flags & FLAG_LOCK) {
		lock_user(fpasswd, username);
	} else if (flags & FLAG_UNLOCK) {
		unlock_user(fpasswd, username);
	} else if (flags & FLAG_DELETE) {
		delete_user(fpasswd, username);
	} else { /* set password */

		if (isatty(STDIN_FILENO)) {
			char* p2;

			passwd = getpass("Enter password: ");
			if (passwd == NULL) {
				fprintf(stderr, "Please specify a password\n");
				exit(1);
			}


			p2 = strdup(passwd);
			passwd = getpass("Re-enter password: ");
			if (passwd == NULL) {
				fprintf(stderr, "Please specify a password\n");
				exit(1);
			}

			if (p2 == NULL || strcmp(passwd, p2) != 0) {
				fprintf(stderr, "Passwords do not match\n");
				exit(1);
			}
			free(p2);
		} else {
			passwd = NULL;
			l = getline(&passwd, &i, stdin);
			if (l <= 1) {
				fprintf(stderr, "Please specify a password\n");
				exit(1);
			}

			free_passwd = 1;
			if (passwd[l-1] == '\n')
				passwd[l-1] = 0;
		}

		crypt_int(fpasswd, username, groupname, passwd);
		if (free_passwd)
			free(passwd);
	}

	free(fpasswd);
	free(groupname);
	gnutls_global_deinit();
	return 0;
}

